Copy geometry when creating data container

hrobarts opened the pull request:
## Description
Creating an AcquisitionData/ImageData with a different dtype will override the geometry,
- Update the docstrings to make this clear

However there is a bug that creating a DataContainer can silently change the dtype of an existing geometry, like in #1952 
- Update code to create AcquistionData or ImageData with a copy of the geometry that's passed

Also clarify precedence for the 3 ways to pass dtype to acq/img data init
	- Array dtype
	- Geometry dtype
	- Kwarg dtype

Behaviour
- If you pass both an array and a geometry, the array dtype will over-ride the geometry dtype
- If you pass just a geometry, the array will be created with the geometry dtype. _This is a change of behaviour, previously it would default to float32_
- Say in the docstring that kwarg dtype is useful if you pass None to array and want to over-ride the dtype of the geometry.
- If you pass kwarg dtype AND an array, it must have the same dtype otherwise an error will be raised. _This is a change in behaviour, previously it would have ignored the kwarg dtype_

Updated docstrings:

![image](https://github.com/user-attachments/assets/aa2582c2-193f-4893-b07c-118dc3a770e5)
![image](https://github.com/user-attachments/assets/1c07c674-d45e-4674-817c-29099318f107)


Questions about VectorData
- VectorData does currently allow no geometry to be passed, so it's a kwarg
- VectorData has deep_copy fixed as True




## Testing you performed
> Please add any demo scripts to https://github.com/TomographicImaging/CIL-Demos/tree/main/misc


## Related issues/links
Closes #1952 

## Checklist

- [x] I have performed a self-review of my code
- [x] I have added docstrings in line with the guidance in the developer guide
- [x] I have updated the relevant documentation
- [x] I have implemented unit tests that cover any new or modified functionality
- [x] CHANGELOG.md has been updated with any functionality change
- [x] Request review from all relevant developers
- [x] Change pull request label to 'Waiting for review'

## Contribution Notes

Please read and adhere to the [developer guide](https://tomographicimaging.github.io/CIL/nightly/developer_guide/) and local patterns and conventions.

- [x] The content of this Pull Request (the Contribution) is intentionally submitted for inclusion in CIL (the Work) under the terms and conditions of the [Apache-2.0](https://www.apache.org/licenses/LICENSE-2.0) License
- [x] I confirm that the contribution does not violate any intellectual property rights of third parties

lauramurgatroyd commented on Wrappers/Python/cil/framework/acquisition_data.py:
```diff
@@ -23,7 +23,18 @@
+    
+    Parameters
+    ----------
+    array : numpy.ndarray or DataContainer
+        The data array.
```
could add that default is empty array of size determined by the geometry?

lauramurgatroyd commented on Wrappers/Python/cil/framework/image_data.py:
```diff
@@ -21,7 +21,19 @@
+    
+    Parameters
+    ----------
+    array : numpy.ndarray or DataContainer
```
could add that default is empty array of size determined by the geometry?

lauramurgatroyd reviewed (changes requested):
Looks good! Could we add a unit test produced from the example of where you had failing code before this change?

hrobarts commented:
After discussion in CIL dev meeting
- Test what happens if we make it fail when geometry.dtype != array.dtype

hrobarts commented:
> After discussion in CIL dev meeting
> 
> * Test what happens if we make it fail when geometry.dtype != array.dtype

A few tests fail, but could be fixed. Specifically it's not obvious how we'd set the acquisition geometry dtype if you create it with one of our factory methods, it needs to be updated after the geometry is created:
```
AG = AcquisitionGeometry.create_Parallel2D()
AG.dtype = np.float64
```
Encouraging setting the geometry dtype could be another way of getting the dtypes out of sync.

Comment from Gemma
The only way I can see to enforce consistency it is to separate geometry and datacotainer objects (i.e. no ImageData/AcqData), and make the user pass both (which often is the case anyway in algorithms/functions/processors) and everything will have to check for shape/dtype-consistency between them where it matters. Or only have the geometry as the base class, which holds the metadata it needs plus the array.

hrobarts commented:
> Looks good! Could we add a unit test produced from the example of where you had failing code before this change?

Thank you for the review Laura, I've updated everything and added a tested from the example.

gfardell reviewed (approved):
Looks good!
